torch/lib/c10d/ProcessGroupMPI.cpp

```diff
@@ -293,4 +309,200 @@ std::shared_ptr<ProcessGroup::Work> ProcessGroupMPI::allreduce(
   return enqueue(std::move(entry));
 }
 
+std::shared_ptr<ProcessGroup::Work> ProcessGroupMPI::reduce(
+    std::vector<at::Tensor>& tensors,
+    const ReduceOptions& opts) {
+  checkSingleTensor(tensors);
+
+  std::function<void(std::unique_ptr<WorkEntry>&)> runFunc =
+      [opts, this](std::unique_ptr<WorkEntry>& entry) {
+        auto data = (*entry->src)[0];
```

Comment:
Do we still do buffering in case of MPI? We really shouldn't because it can work directly on the tensor data. Otherwise it will be a perf regression compared to THD.

Reply:
We don't,  we just pass the data pointer to the worker thread